docs/proposals/api-group.md

```diff
@@ -0,0 +1,150 @@
+
+    TypeMeta is an exception, because it is a common object that is used by objects in all groups but does not logically belong to any group. We plan to move it to the package `unversioned`.
+
+## Client-side implementation
+
+1. clients:
+
+  Currently we have structured (pkg/client/unversioned#ExperimentalClient, pkg/client/unversioned#Client) and unstructured (pkg/kubectl/resource#Helper) clients. The structured clients are not scalable because each of them implements specific interface, e.g., [here](../../pkg/client/unversioned/client.go#L32). Only the unstructured clients are scalable. We should either auto-generate the code for structured clients or migrate to use the unstructured clients as much as possible.
+
+2. Spelling the URL:
+
+  The URL is in the form of prefix/group/version/. The prefix is hard-coded in the client/unversioned.Config (see [here](../../pkg/client/unversioned/experimental.go#L101)). The client should be able to figure out `group` and `version` using the RESTMapper. For a third-party client which does not have access to the RESTMapper, it should discover the mapping of `group`, `version` and `kind` by querying the server as described in point 2 of #server-side-implementation.
+
+3. kubectl:
+
+  kubectl should accept arguments like `group/resource`, `group/resource/name`. Nevertheless, the user can omit the `group`, then kubectl shall rely on RESTMapper.VersionAndKindForResource() to figure out which group the resource belongs to, or return an error if there is ambiguity. A special case is when a resource (like `node`) exist in both k8s v1 API and k8s modularized API (like `infra/v2`), kubectl should use the modularized API rather than returning an error because of the ambiguity.
+
+  The --api-version flag of kubectl should accept values in the form of `group/version`.
+
+## OpenShift integration
+
+OpenShift can take a similar approach to break monolithic v1 API: keeping the v1 where they are, and gradually adding groups.
+
+For the v1 objects in OpenShift, they should keep doing what they do now: they should remain registered to Scheme.versionMap["v1"] scheme, they should keep being added to originMapper.
+
+For new OpenShift groups, they should do the same as native Kubernetes groups would do: each group should register to Scheme.versionMap["group/version"], each should has separate RESTMapper and the register the MultiRESTMapper.
+
+To expose a list of the supported Openshift groups to clients, OpenShift just has to call to pkg/cmd/server/origin#call initAPIVersionRoute() as it does now, passing in the supported "group/versions" instead of "versions".
+
+
+## Future work
+
+1. Dependencies between groups: we need an interface to register the dependencies between groups. It is not our priority now as the use cases are not clear yet.
```

Comment:
As API groups get more granular, cross-group references seem very likely. I think it would be worthwhile to think though how we would define/migrate (and keep valid as servers eventually disable older API versions) those "fully qualified" stored references.